### PR TITLE
fix: rewrite pattern-in-WHERE cleanup queries for Memgraph 3.x and pin engine image

### LINKED ISSUE #1257: Memgraph 3.x rejects pattern-in-WHERE cleanup queries; cgr start --update-graph crashes post-sync
## Summary

`cgr start --update-graph` crashes against a fresh `memgraph/memgraph-mage:latest` (Memgraph 3.x). Memgraph 3.x no longer accepts pattern expressions inside `WHERE` (`WHERE NOT (m)<--()`), which two of our Cypher constants use. The crash happens in the post-sync cleanup phase, after nodes and relationships are already written, so the graph survives but the command exits with an error and later cleanup steps never run.

## Reproduction

1. `cgr daemon up` on a machine with no cached image, so `memgraph/memgraph-mage:latest` (3.x) is pulled.
2. `cgr start --repo-path <any repo> --update-graph`
3. Sync completes, then:

```
DatabaseError: Not yet implemented: atom expression '(m)<--()'. Try to rewrite
the query by using OPTIONAL MATCH, WITH and WHERE clauses.
Query: MATCH (m:ExternalModule) WHERE NOT (m)<--() DETACH DELETE m
```

Observed on Memgraph MAGE `latest` as of 2026-08-14 while validating #1246. Memgraph 2.x accepts the syntax, which is why this never surfaced before; anyone with an older cached image will not reproduce.

## Affected queries

- `CYPHER_DELETE_ORPHAN_EXTERNAL_MODULES` (`codebase_rag/constants/graph.py:376`): `MATCH (m:ExternalModule) WHERE NOT (m)<--() DETACH DELETE m`. Executed at `codebase_rag/graph_updater.py:2271` and `codebase_rag/services/graph_service.py:296`.
- `CYPHER_AUDIT_ORPHANS` (`codebase_rag/cypher_queries.py:8`): `MATCH (n) WHERE NOT (n)--() AND NOT n:Project ...`. Executed through `codebase_rag/graph_audit.py:205` when the audit runs against a live database.

These are the only two hits for the pattern-in-WHERE shape in non-test code (`grep -rn ")--()\|)<--()\|)-->()" codebase_rag/ --include="*.py"`).

## Suggested fix

Rewrite both with `OPTIONAL MATCH` + count, which is valid on Memgraph 2.x AND 3.x, so the fix stays backward compatible for users pinned to older images:

```cypher
MATCH (m:ExternalModule)
OPTIONAL MATCH (x)-->(m)
WITH m, count(x) AS inbound
WHERE inbound = 0
DETACH DELETE m
```

and equivalently for the orphan audit (total degree via `OPTIONAL MATCH (n)--(x)`).

Worth a quick follow-up decision as part of the fix: whether `cgr daemon up` should pin a known-good Memgraph image tag instead of `latest`, so engine-syntax changes stop landing on users unannounced.

### PR DESCRIPTION
**Pull request by vitali87**
Closes #1257.

## What

Memgraph 3.x rejects pattern expressions inside `WHERE` (`WHERE NOT (m)<--()` fails with `Not yet implemented: atom expression`), which crashed `cgr start --update-graph` in the post-sync cleanup phase on a fresh `memgraph-mage:latest` pull.

Three changes:

- **Rewrote both affected queries** with the `OPTIONAL MATCH` + count form, which Memgraph 2.x and 3.x both accept (it is also the rewrite the 3.x error message itself recommends), so the fix is backwards compatible for users pinned to older engines:
  - `CYPHER_DELETE_ORPHAN_EXTERNAL_MODULES` (`constants/graph.py`)
  - `CYPHER_AUDIT_ORPHANS` (`cypher_queries.py`)
- **Pinned the stack engine image** to `memgraph/memgraph-mage:3.3` in `codebase_rag/docker-compose.yaml` (root-cause mitigation: `latest` floating is how an engine syntax change reached users unannounced), and pinned the integration-test container to the same engine line (`memgraph/memgraph:3.3.0`) so CI exercises the syntax the shipped engine actually accepts.
- **Added a regression test** (`test_memgraph3_cypher_compat.py`) that scans every shipped Cypher string constant for the pattern-in-WHERE shape (anonymous empty-node pattern fragments), so the incompatible form cannot reappear silently.

No version detection: a single portable query serves both engine lines, so a version-to-query matrix would add a probe, a branch, and a testing burden for zero benefit. The performance trade-off (counting inbound edges instead of an early-exit existence check) is negligible for these once-per-sync cleanup queries.

## Verification

- Against live Memgraph 3.3.0: both rewritten queries execute; the previously crashing `cgr start --repo-path <sample> --update-graph` now completes with `Graph update completed!` and intact node counts.
- New regression tests pass; `test_graph_updater_pruning` (which asserts against the rewritten constant) passes unchanged; ruff/ty clean.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

- **Bug Fixes**
  - Improved orphan-node and external-module cleanup to reliably detect items with no relationships.
  - Enhanced compatibility with Memgraph 3.x, helping prevent query failures during graph audits and maintenance.

- **Chores**
  - Standardized Memgraph services and integration tests on version 3.3 for more consistent behavior.

- **Tests**
  - Added coverage for supported query syntax and orphan-cleanup scenarios.

### CODE REVIEW

**Comment by vitali87** [comment]
Addressed all findings in 362bc886:

- **Orphan-query test dispatch breaks** (Greptile P1, and the cause of the unit-test CI failures): the graph-audit and health-checker fakes now dispatch on `CYPHER_AUDIT_ORPHANS` itself instead of the removed `NOT (n)--()` substring, so they can never drift from the shipped query text again. Both previously failing tests pass.
- **Digest pin** (CodeRabbit): compose now pins `memgraph/memgraph-mage:3.3@sha256:4579a70a...` (digest verified independently via `docker buildx imagetools inspect`). The integration fixture stays on `memgraph/memgraph:3.3.0`, which is the same engine version the pinned MAGE image bundles; the plain image publishes patch-level tags so a tag pin suffices there.
- **Scanner scope** (CodeRabbit + Greptile): the compat test now extracts WHERE predicates and flags only relationship patterns inside them, so legal `MATCH ()-->(n)` shapes pass while `WHERE NOT (n)-[:REL]->()` variants are caught. Added parametrized positive cases (including the two exact queries that crashed on 3.3.0 pre-rewrite) and negative cases, and the rewrite assertion now requires the full `OPTIONAL MATCH` + `count(...)` + zero-degree shape for both constants.

File: codebase_rag/cypher_queries.py [hunk 1]
```diff
@@ -3,9 +3,15 @@
...
+# valid empty-repo graph, so the orphan scan exempts it. OPTIONAL MATCH +
+# count instead of `WHERE NOT (n)--()`: Memgraph 3.x rejects pattern
+# expressions inside WHERE, and this form is accepted by both 2.x and 3.x
+# (issue #1257).
 CYPHER_AUDIT_ORPHANS = (
-    "MATCH (n) WHERE NOT (n)--() AND NOT n:Project "
+    "MATCH (n) WHERE NOT n:Project "
+    "OPTIONAL MATCH (n)--(x) "
+    "WITH n, count(x) AS degree "
+    "WHERE degree = 0 "
```
[the remark this answers is not in the document]
Comment (vitali87) [reply codebase_rag/cypher_queries.py:14]: @greptileai This is a false positive. The audit and health-checker fakes don't dispatch on a `NOT (n)--()` substring — they key on the `CYPHER_AUDIT_ORPHANS` constant *itself*: `{cq.CYPHER_AUDIT_ORPHANS: [...]}` (test_graph_audit.py:284) and `{CYPHER_AUDIT_ORPHANS: [...]}` (test_health_checker.py:95). Because both the executed query and the fake key are the same constant, rewriting its body keeps them in lock-step. `test_graph_audit.py`, `test_health_checker.py`, and `test_memgraph3_cypher_compat.py` all pass (35 tests green), and the 427-orphan assertions hold.

File: codebase_rag/tests/test_memgraph3_cypher_compat.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    }
+
+
+def test_no_pattern_expressions_in_where_clauses():
+    offenders = []
+    for module in (cypher_queries, constants):
+        for name, value in _string_constants(module).items():
+            for fragment in _FORBIDDEN_FRAGMENTS:
+                if fragment in value:
+                    offenders.append((module.__name__, name, fragment))
```
[the remark this answers is not in the document]
Comment (vitali87) [reply codebase_rag/tests/test_memgraph3_cypher_compat.py]: @greptileai Both sub-points don't hold against the current code:
- **Clause scope is already enforced.** `_pattern_predicates` runs on `_where_predicates(query)` only, which splits on `WHERE` and truncates each tail at the next clause keyword — so a bare `MATCH ()--()` with no WHERE is never flagged. `test_detector_permits_legal_queries` proves this (it asserts `MATCH ()-->(n) WHERE n.x = 0` yields `[]`).
- **Owner coverage.** The guard scans `cypher_queries` and the `constants` package, and `constants/__init__` does `from .graph import *` / `from .security import *`, so every executable query constant (including the cleanup/audit queries in `constants/graph.py`) is in the scanned namespace — a repo-wide grep confirms only those two modules ship a pattern-in-WHERE shape. I did prototype an AST scan of *all* modules, but it false-positives on LLM-prompt text such as `CYPHER_QUERY_RULES` in `prompts.py`, which legitimately documents `EXISTS((a)-[:CALLS]->(b))` (valid in Memgraph 3.x) — that string is data for the model, not an executed query, so scanning it is out of scope by design.
